[Time Conductor] #933 Time Conductor V2 API proposal

akhenry opened the pull request:
Initial pass at a notional time conductor v2 API.

See README for details and to make comments inline.

VWoeltjen commented on platform/features/conductor-redux/src/README.md:
```diff
@@ -0,0 +1,187 @@
+with inner interval unlocked. Not sure what result would be, has inner end switched to fixed?
+Also example on page 55 in real-time where inner end < now. Is there a use case for this? Semantically, it's saying
+show me real time, but stale data. Why would a user want this? My feeling is that if the inner 
+OR outer ends are moved behind NOW in real-time mode then you drop into historical mode.
```
We discussed this a bit on Friday, and were in overall agreement that there wasn't a use case for end < now with RT. Some alternatives were discussed (I think your proposal to drop into historical mode was among them) but no final decision made. In any event, don't think this has any impact on the API

VWoeltjen commented on platform/features/conductor-redux/src/README.md:
```diff
@@ -0,0 +1,187 @@
+* Have not included validation (eg. start time < end time) or any other 
+business logic such as what happens when outer interval gets dragged 
+within range of inner interval. Focus is on teasing out the public API 
+right now. 
```
I do think this is an API concern - if I set the end before the start, what happens? (Exception? Value gets clamped? New value is just accepted?) But yes, do agree that it is better to work through the nominal cases first.

akhenry replied:
Agree, validation should happen within the TimeConductor itself. Wanted to establish the functional API first, then think about validation concerns. There may be some value in exposing validation functions from the TimeConductor API to avoid duplicating validation in several places (in the TimeConductor class, as well as in the API).

VWoeltjen commented on platform/features/conductor-redux/src/README.md:
```diff
@@ -0,0 +1,187 @@
+
+### Additional possible use-cases
+1. Telemetry adapter wants to indicate presence of data at a particular time
+2. Time conductor controller wants to paint map of data availability.
```
Agree with omitting these use cases. This information is useful visually but not relevant to the information the time conductor is modeling (user's desired time frame and settings.)

VWoeltjen commented on platform/features/conductor-redux/src/README.md:
```diff
@@ -0,0 +1,187 @@
+5. Switch to follow time. Set inner end to now, and outer end to now + 1 hour
+6. Plot controller updated on tick
+7. Plot Controller updated when user changes bounds
+8. Conductor controller needs to update bounds and mode on TC when user changes bounds
```
Think these are a good list of use cases. Notably, I think the difference between Use Case 6 and 7 is that the plot will want to query for new data in the case of 7, but not in the case of 6.

VWoeltjen commented on platform/features/conductor-redux/src/README.md:
```diff
@@ -0,0 +1,187 @@
+    tc.bounds.inner.start.time(session.metadata.firstTime);
+    tc.bounds.inner.end.time(session.metadata.lastTime);
+    tc.bounds.outer.start.time(session.metadata.firstTime);
+    tc.bounds.outer.end.time(session.metadata.lastTime);
```
Couple of things here:

* I don't believe that "outer" bounds need to be expressed as part of the time conductor at this level (the model). It's an important part of the visual control, but the information that matters to the system is the inner bounds. (Credit to @larkin for making that point to me when I was initially implementing the TC.)
* I'd also like to reduce interface depth here; this reads reasonably as an example, but starting at `tc` and getting to `tc.bounds.inner.start.time` using API docs would be a bit of an exercise.

To address both those points, I'd suggest just having a `tc.start(timestamp)` and `tc.end(timestamp)`

akhenry replied:
Fair enough. This is another case of separating the public API from the code used by the time conductor UI component. I'll take another look at it.

VWoeltjen commented on platform/features/conductor-redux/src/README.md:
```diff
@@ -0,0 +1,187 @@
+    
+    //Set start and end modes to fixed date
+    tc.startMode(tc.Modes.FIXED);
+    tc.endMode(tc.Modes.FIXED);
```
The use cases involving modes here are all initiated from the time conductor in platform; we don't have cases where, say, a plugin would want to set this. Because of that, I'm inclined to remove modes from the public API, and (corollary to that) to handle them architecturally with a separate component from `MCT.conductor`. The same UI component would modify the strategy this component was using, and this component would update the TC start/end times in appropriate ways at appropriate times.

This would reduce the scope of the public API, and narrow the concerns of the conductor. It does introduce some challenges in differentiating between Use Cases 6 & 7 above; the start/end setters&ast; could take an `options` argument with a flag to distinguish these.

&ast; _There's a good chance we'll also want a `bounds(start, end, [options])`, particularly if `start`/`end` calls are sensitive to ordering; depends on our validation decision._

akhenry replied:
I take your point. I will look at refactoring to better separate the public-facing API from implementation details like this. The modality of the TimeConductor probably doesn't need to be exposed at this level of detail, but there may be value in exposing a function that allows plugins to switch between historical and real-time conductor modes.

VWoeltjen commented on platform/features/conductor-redux/src/README.md:
```diff
@@ -0,0 +1,187 @@
+    
+    // Delta defined in ms. Start time delta is relative to end time.
+    tc.bounds.inner.start.delta(FIFTEEN_MINUTES);
+    tc.bounds.outer.start.delta(FIFTEEN_MINUTES);
```
I think this `delta` parameter could be removed, and the responsibility for converting from deltas to bounds delegated to the Mode:

```
tc.timeSystem(telemetryMetadata.timeSystem);
tcMode.switch(new FollowMode(FIFTEEN_MINUTES));
```

Again, idea here is to simplify public API - no need to worry about which mode you're in, you just have the ability to observe the parameters (start, end, TOI, time sys) which will effect queries or display elsewhere in the UI.

VWoeltjen commented on platform/features/conductor-redux/src/README.md:
```diff
@@ -0,0 +1,187 @@
+// handles being dragged) only.
+MCT.conductor.bounds.inner.listen(function(bounds) {
+    plotUpdater.setDomainBounds(bounds.start.time(), bounds.end.time());
+}, MCT.conductor.bounds.EventTypes.USER);
```
Not to stray too far out of scope, but this looks like a suitable opportunity to make a broader design decision around change notification. This appears to extend our current approach, which I am a little hesitant to do when there are third-party options available. Additionally, looking at the API redesign there are a lot of cases where we'll want to provide some form of change notification for objects (especially true as we rely less on Angular and the digest cycle to detect changes for us)

There are three paths I see as viable here:

1. [EventEmitter](https://nodejs.org/api/events.html) (or, more specifically, one of the browser implementations)
  * Pros:
    * Very standard; since this API is built into NodeJS it will seem native to some developers.
    * Multiple implementations are available to pick and choose from.
    * JSDoc has a tag for [events](http://usejsdoc.org/tags-event.html), making this fairly documentable
    * Allows the same object to fire multiple event types (as is done with this `listen` call)
  * Cons:
    * Based on string-matching, which we've generally wanted to avoid. (But, one of the detriments of string-matching based approaches is the difficulty in documenting them, which JSDoc support mitigates)
    * More methods are exposed than we might generally wish to (and some implementations have even more.) In particular, `emit` is exposed everywhere.
2. Use observables/reactive programming (e.g. [Kefir](https://rpominov.github.io/kefir), [RxJS](https://github.com/Reactive-Extensions/RxJS), [Bacon](https://baconjs.github.io/))
  * Pro: Include support for chaining observables in various ways (maps, filters, etc), which could greatly simplify the backend-to-view data flow.
  * Con: Non-standard, and tend to expose rather large APIs. Hard to change this decision later, esp. if we expose these as part of our public API (and listening for changes is a public API use case)
3. Extend our existing event system
  * Pros: We get exactly what we want and nothing more.
  * Cons: More effort, testing, maintenance required.

Of the options, EventEmitter certainly looks like the safest choice. I think the other options are "better" in certain respects (reactive programming is powerful, and having a small API that does exactly what we need is great, too) but it would be hard to argue that they're better enough to justify the investment/risk

akhenry replied:
@VWoeltjen Good points. Yes, from the outset I made the assumption that we would be using our current observable pattern approach, without really considering alternatives. You're right, this seems like a good time to standardize our approach here for the broader API.

I did have an earlier iteration that was quasi-EventEmitter like but with enum matching instead of string matching. I convinced myself that it was simpler and clearer to just attach listeners to the object you're interested in (eg. the inner bounds) than to specify strings or enum values for the event type you're interested in, but I'm not so sure about it and I keep wavering. Having a single listener attachment point on the TimeConductor does make a lot of sense, and removes any ambiguity about which object to listen to. Being able to properly document this with JSDoc will increase clarity about event names.

Can see advantages to event streams for continuous events like ticks.

Tend to agree with your assessment. EventEmitter sounds like a good approach.

larkin commented:
Some agreed-upons:
* Yes to eventemitter + string names for events.  It's simple, performant, external developers can provide adaptations (or wrappers) for other patterns.
* Yes to removing concept of "outer bounds" from TC API.

I think we need to batch updates into meaningful and valid chunks; meanwhile, 
there should only be a few types of updates a view should be expected to handle:

    * `change` when everything changes, e.g. when time system changes.
    * `change:toi`: Update TOI.
    * `change:bounds`: update both start and end (usually from tick)
    * `change:bounds:end`: update only end (usually from tick)
    * `change:bounds:start`: update only start (usually from tick)
    * `synchronize`: views may ignore `changes`.  views should not ignore 
        `synchronize`.

Question: We could remove `change:bounds:end` and `change:bounds:start`-- or is 
that helpful to views?

**Validation**: I think TC API throws errors if you give it invalid values-- it 
just blows up and throws an exception.  Views should be sanitizing user inputs 
before sending them to the TC API.

**Time Systems**: TC API should be responsible for broadcasting changes based
on time system changes.  May require certain configuration to work, e.g.
```javascript
    tc({
        system: UTCTimeSystem
        start: {
            value: ONE_HOUR
            mode: tc.Modes.DELTA,
        },
        end: {
            mode: tc.Modes.RELATIVE,
            value: UTCTimeSystem.now
        },
        toi: {
            mode: tc.Modes.RELATIVE,
            value: UTCTimeSystem.now
        }
    });
```

I wonder if we want to expose a simple set of objects for specifying times, e.g.

```javascript
new tc.time.RELATIVE({
    tickSource: somethingThatTicks,
    offset: '4 hours',
});

new tc.Time.FIXED({
    value: blah.
});
```

akhenry commented:
Yyyyyyeeeeeeahhhhh. I'm not wild about the string matching, especially with the relative complexity of those strings. It suggests that `EventEmitter` may not necessarily be the best fit here. It's great for objects that emit a lot of different types of events, but our time conductor mostly just emits property change events, and we're interested in the properties that have changed.

A custom change handler might be a better fit here. We could roll our own that sticks closely to the EventEmitter interface -

``` javascript
tc.on('change', function(tc){...});
tc.on('change', 'toi', function(toi){...});
tc.on('change', 'bounds', function(bounds){...});
tc.on('synchronize', function(tc){...});
```
If we're not actually interested in the property that has changed (and I'm not convinced we are in most cases), then we could use `EventEmitter` and simplify it like so
``` javascript
tc.on('change', function(tc){...});
tc.on('bounds', function(bounds){...});
tc.on('synchronize', function(tc){...});
```
We could optionally pass the name of the property that changed in the callback to allow for functions that need to differentiate. So long as 'bounds' events and tc change events are separated we won't get a ton of unnecessary callbacks.
``` javascript
tc.on('change', function(tc, property){...});
tc.on('bounds', function(bounds, property){...});
tc.on('synchronize', function(tc){...});
```

As a side note, I don't know what `synchronize` does?

> Question: We could remove change:bounds:end and change:bounds:start-- or is 
that helpful to views?

Agreed. I'm not sure it's necessary to differentiate. Suggest we leave it out and see if it crops up as a need in implementation.

> Validation: I think TC API throws errors if you give it invalid values

Yep, agree. UI will also need to do separate validation, but the public facing API should throw errors on invalid values.

> ``` javascript 
new tc.time.RELATIVE({
    tickSource: somethingThatTicks,
    offset: '4 hours',
});
```
Yeah I was thinking of externalizing the ticking, but I'm not really sure why :) Might be useful different playback rates?

I also want to go back to the need to differentiate between ticks and user bounds changes. Do we really need to do this in the public API?

larkin commented:
Those strings are pretty basic-- they just communicate what could otherwise be a constant.  But they can be more clear.  Keep in mind strings are interned in V8 so equality comparisons are quite performant.  If it's worth communicating more information to a developer in the string, we should do it.  That said, removing the prefixes seems like a good way to simplify.   after that, there are really only three events worth listening to:

`change`: expect everything changed.  update accordingly. (this is a relative rare event to trigger, used to prevent racy behavior by updating everything at once)
`toi`: toi has changed.  update accordingly. (this is probably the noisiest event to trigger as it is frequently tied to noisy input events like mouseover)
`bounds`: only bounds has changed.  update accordingly. (this is a common event, happens on ticks, but not as bad as `toi`).

We don't want to place the burden of dirty checking on views-- events should be granular enough to allow them to update as needed without having to debounce everything.

as far as `synchronize`: If we don't separate user inputs from automatic updates, we need some way of forcing that rebellious zoomed plot view back into order-- and that's `synchronize`.  Fits with some design goals of having a gesture that forces all views to synchronize with the time conductor.  Though, we don't need `synchronize` up front, my feeling is `change` will be good enough.

IMO, The time conductor api should not generate ticks-- it should listen on another thing which will trigger tick events.  The control flow is important-- the time conductor isn't ticking, the telemetry source / time system / utc clock / time bomb is ticking, and the time conductor is broadcasting that event.

VWoeltjen commented:
Don't want to add too much noise, but agree that we can get this down to around three events. Maybe two if we pass meaningful event objects. I'm not going to look at names, just want to focus on the set of things we're naming. Primarily I want to tie these back to use cases.

1. An event that signifies a bounds and/or time system change (user-initiated or otherwise)
  * Use case: Plot wants to keep its X axis in sync with the time conductor. It can listen to this event to do so. It does not worry about requerying on this event. 
2. An event that signifies that any data associated with previous bounds has become obsolete.
  * Use case: Plot (and other views) want to requery when necessary. In the case of Plot, it doesn't want to requery on every bounds change. 
  * In the case of Table, it may not care about the bounds at all (and if it does, it still only wants to requery on user gestures, not ticks)
3. An event that signifies that the time of interest has changed.
  * Use case: LAD-style views, which want to requery on this event specifically.
  * Use case: Timeline, which will want to show some kind of synchronized marker.

@larkin - I think this is highly similar to what you've stated, except that Event 2 ("synchronize") is more clearly distinguished from Event 1. Has nothing to do with bounds changes; when a user changes bounds, you get Event 1 followed by Event 2.

Event 2 could be removed if some appropriate flag was passed in on an event object from Event 1, but just separating the events is probably easier to use.

For Event 3, it's worth considering whether time of interest is really "part" of the time conductor (at the level of the data model), or if it's a related object that is operated by the same visual control. Conceptually, I think it does make sense to separate them (Law of Demeter: why should I need a reference to the whole time conductor if I am _only_ interested in the TOI?) But I also see the value of saying "I see this thing in the UI, here's the object in memory that represents its state."

What does the plot say?

```js
tc.on('change', updateDisplayBounds);    // Event 1
tc.on('synchronize', requery);           // Event 2
tc.on('toi', updateTimeOfInterestLine);  // Event 3
```

> IMO, The time conductor api should not generate ticks-- it should listen on another thing which will trigger tick events. The control flow is important-- the time conductor isn't ticking, the telemetry source / time system / utc clock / time bomb is ticking, and the time conductor is broadcasting that event.

Agree with this. Tick source ticks -> mode updates TC bounds -> TC fires event.

akhenry commented:
## Update

Have addressed comments from code review
* Bounds have been radically simplified (as has api generally)
* Focused on what's required in the public API. Some aspects of the original API proposal were addressing functionality of the Time Conductor UI, which are no longer addressed here.
* Events simplified:
    * bounds - start, end, or both have changed
    * refresh - Something has changed that means the whole thing should be re-queried and the view refreshed. Now, importantly, I have included user bounds changes (bringing those rogue zoomed plots back into line) and time system changes here.
    * timeOfInterest - time of interest has changed.
* Updated the example code in the use cases
* Externalized modes, and ticking (not to be confused with visual ticks :) ). Introduced notion of TimingSource (something that ticks, could be local clock for real-time, a remote timing source, or a telemetry source that ticks on LAD)

larkin commented:
notes from long discussion

```javascript


/*
 events:
'bounds'
'follow'
'system'
'timeOfInterest'

note: A 'system' event is always followed by a 'bounds' event.
Or more precisely, once a system event has been received, views should wait for
a bounds event before updating their view.  It's probably a good time to show a
loading indicator.

Can we assume that follow "true" + bounds = you should have received data via
subscription?  Nope, but for now, for most future cases, we can.  We know
predicted data that is stored as historic data will cause issues but we think
that is better addressed in the future.

We see some future need for a 'synchronize' to force views to synchronize, but
no need to implement now.  YAGNI!

Tick sources-- tick sources will not be a part of the formal API, time systems
should define tick sources but the time conductor API will not expose a method
for picking which tick source is used.  At this time, the Time Conductor View
will have this sole responsibility.  If it is necessary to add it to the public
API later, we can do that.

*/

tc.on('bounds', plot.panIfNotDisconnected);

tc.on('follow', plot.warnIfFollowingAndDisconnected);

tc.on('system', plot.setSystemAndReconnect);

tc.on('timeOfInterest', plot.highlightTOI);

plot.focusArea = function (start, end) {
    MCT.conductor.follow(false);
    MCT.conductor.bound({start: start, end: end});
}

// YAGNI, so far.  we think we need it, we can wait for a bug report later.
// tc.on('synchronize', plot.reconnect);
```

akhenry commented:
Closed as time conductor will be merged as part of https://github.com/nasa/openmct/pull/1089
